src/mongo_odm/odm_collection.hpp

```diff
@@ -0,0 +1,259 @@
+    }
+
+    ///
+    /// Runs an aggregation framework pipeline against this collection, and returns the results
+    /// as
+    /// de-serialized object.
+    ///
+    /// @param pipeline
+    ///   The pipeline of aggregation operations to perform.
+    /// @param options
+    ///   Optional arguments, see mongocxx::mongocxx::options::aggregate.
+    ///
+    /// @return A deserializing_cursor with the results.
+    /// @throws
+    ///   If the operation failed, the returned cursor will throw an exception::query
+    ///   when it is iterated.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/command/aggregate/
+    ///
+    deserializing_cursor<T> aggregate(
+        const mongocxx::pipeline& pipeline,
+        const mongocxx::options::aggregate& options = mongocxx::options::aggregate()) {
+        return deserializing_cursor<T>(_coll.aggregate(pipeline, options));
+    }
+
+    ///
+    /// Finds the documents in this collection which match the provided filter.
+    ///
+    /// @param filter
+    ///   Document view representing a document that should match the query.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::find
+    ///
+    /// @return Cursor with deserialized objects from the collection.
+    /// @throws
+    ///   If the find failed, the returned cursor will throw exception::query when it
+    ///   is iterated.
+    ///
+    /// @see http://docs.mongodb.org/manual/core/read-operations-introduction/
+    ///
+    deserializing_cursor<T> find(
+        bsoncxx::document::view_or_value filter,
+        const mongocxx::options::find& options = mongocxx::options::find()) {
+        return deserializing_cursor<T>(_coll.find(filter, options));
+    }
+
+    ///
+    /// Finds a single document in this collection that match the provided filter.
+    ///
+    /// @param filter
+    ///   Document view representing a document that should match the query.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::find
+    ///
+    /// @return An optional object that matched the filter.
+    /// @throws exception::query if the operation fails.
+    ///
+    /// @see http://docs.mongodb.org/manual/core/read-operations-introduction/
+    ///
+    mongocxx::stdx::optional<T> find_one(
+        bsoncxx::document::view_or_value filter,
+        const mongocxx::options::find& options = mongocxx::options::find()) {
+        return to_optional_obj<T>(_coll.find_one(filter, options));
+    }
+
+    ///
+    /// Finds a single document matching the filter, deletes it, and returns the original as a
+    /// deserialized object.
+    ///
+    /// @param filter
+    ///   Document view representing a document that should be deleted.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::find_one_and_delete
+    ///
+    /// @return The deserialized object that was deleted from the database.
+    /// @throws exception::write if the operation fails.
```

Comment:
It might be worth noting that the exceptions thrown here are all `mongocxx::exception::...` exceptions, and not `mongo_odm::`.